src/core/thread/anycast_locator.cpp

```diff
@@ -0,0 +1,160 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ *   This file implements Anycast Locator functionality.
+ */
+
+#include "anycast_locator.hpp"
+
+#if OPENTHREAD_CONFIG_TMF_ANYCAST_LOCATOR_ENABLE
+
+#include "common/code_utils.hpp"
+#include "common/instance.hpp"
+#include "common/locator_getters.hpp"
+#include "thread/thread_tlvs.hpp"
+#include "thread/uri_paths.hpp"
+
+namespace ot {
+
+AnycastLocator::AnycastLocator(Instance &aInstance)
+    : InstanceLocator(aInstance)
+    , mCallback(nullptr)
+    , mContext(nullptr)
+    , mAnycastLocate(UriPath::kAnycastLocate, HandleAnycastLocate, this)
+
+{
+    Get<Tmf::Agent>().AddResource(mAnycastLocate);
+}
+
+Error AnycastLocator::Locate(const Ip6::Address &aAnycastAddress, Callback aCallback, void *aContext)
+{
+    Error            error   = kErrorNone;
+    Coap::Message *  message = nullptr;
+    Ip6::MessageInfo messageInfo;
+
+    VerifyOrExit((aCallback != nullptr) && Get<Mle::Mle>().IsAnycastLocator(aAnycastAddress),
```

Comment:
Small question:
Is it useful to use this mechanism to convert RLOC to ML-EID?

As a purpose of this functionality, we want to convert ALOC to ML-EID because ML-EID is more stable. 
The same should apply to RLOC. 

Should we allow converting either ALOC or RLOC to ML-EID ? This enhancement should not add any code size, but may even reduce code size.

Thoughts?

@abtink @librasungirl @jwhui

Reply:
I think RLOC is generally used by the Thread core stack, not upper layer application. 

@simonlingoogle is there use case to associate RLOC with ML-EID in your mind?

Reply:
We can consider using it for RLOC as well but I am not sure if we have use-case for it. I prefer limiting it to `Anycast` for now (unless we have a good reason/use-case to extend it to RLOC).

> As a purpose of this functionality, we want to convert ALOC to ML-EID because ML-EID is more stable.
The same should apply to RLOC.

The main goal is to determine the closest destination of ALOC address (since we can have multiple destinations for the same anycasy address). I don't think it is related to ML-EID being more stable.

When using `ML-EID` as destination of message, device needs to address query to discover the RLOC. So if we have RLOC of destination, it sounds more reasonable to use it directly than to try to get the related ML-EID and then use it instead.

Reply:
> When using `ML-EID` as destination of message, device needs to address query to discover the RLOC. So if we have RLOC of destination, it sounds more reasonable to use it directly than to try to get the related ML-EID and then use it instead.

Considering connection protocols like TCP, it would be better to use more stable addresses like ML-EID or OMR address than to use RLOC, which changes as the topology changes. 

That being said, I am fine to consider RLOC when we have a use-case for it.